### ISSUE TITLE
Protobuf for all types?

### ISSUE THREAD

**Issue by ebuchman** [tendermint/tendermint#218]
I'm looking at using grpc for the rpc server. It would ease client adoption in every language and simplify evolution/maintenance of these apis. Trouble is, types must be protobuf, and our rpc calls return types from all over the stack - for this to make sense, we'd basically have to use protobufs for any and all types that at some point go over a wire (pretty much everything) or else maintain duplicates of each type, which seems demanding and fragile. Using protobufs for all types has the advantage that implementations of the whole stack in other languages becomes a lot easier. Disadvantage is the lock-in and the mess it will make of the code.

**Comment by mdyring** [comment]
I only use gRPC / protobuf at the "edge".. At least on C++ the protobuf generated code is not very pretty, useS Google specific containers etc.

I had the same considerations and resisted, as everything would need to be shoehorned into Protobufs view of the world.

IMO a cleaner approach is just to restrict the Protobuf/gRPC code to the interface to the outside world, and then do some conversion/copying there.

**Comment by ebuchman** [comment]
agreed, but the rpc exposes all kinds of inner types to the outside world. we could copy them all, but then we'd be maintaining two copies of many of the types (ie. since blocks, sigs, peer info, consensus data etc are returned via rpc, nearly all times are implicated, and would need duplication)

**Comment by silasdavis** [comment]
In the past I was tempted to do something similar using Avro-generated types. I lived to regret it, because I found that they changed for different reasons. Sometimes I wanted some book-keeping values for the in-memory version that I didn't want in the serialisation type and conversely sometimes I wanted values only on the serialisation type that were written on finalisation. I ended up having to wrap the Avro types in memory, and I was stuck with the Avro data structures and as @mdyring says the serialisation library's view of the world.

Some examples are maybe wanted a list data structure for serialisation but a map or various other indices in-memory.

Even if it doesn't feel very DRY I would seriously consider maintaining pairs of serialisation types and in-memory types. I haven't looked at much go protobuf code, but what I have seen (like Tendermint types.pb.go) doesn't look particularly nice.
